### ISSUE TITLE
Add a not-executed hook to a solid

### ISSUE THREAD

**Issue by AmirAktify** [dagster-io/dagster#3773]
## Use Case
Sometimes in a DAG, a "finalize" solid become not reachable due to an intermediate pipeline step failing. (So neither "success" or "fail"). In this case, it would be nice to have a "not_reachable" / "not_executed" hook which is triggered, and takes over alerting or resources cleanup.

## Ideas of Implementation
Adding a "not executable / not executed hook", once the executor decides that a solid is going to be skipped / never run. There are probably work arounds for my particular use-case; but in general I think its a nice abstraction to have available when associating processes/flows with a DAG.

## Additional Info

---
#### Message from the maintainers:
Excited about this feature? Give it a :thumbsup:. We factor engagement into prioritization.

**Comment by yuhan** [comment]
Hi @AmirAktify, happy to learn more about your use case. When you say "not-executed" / "skipped", which case(s) are you referring to:
1) a solid was supposed to run but some upstream failure stopped the pipeline process, so the solid never ran.
2) a solid execution was skipped because of conditional execution, i.e. you expect the solid execution to skip.

the proposed solutions could be a bit different
1) the current hook implementation is in-process, so if the pipeline fails, the not-executed solids in this case won't run so neither would their hooks. To tackle this case, we will likely need to implement some out-of-process hook triggering mechanism that will always evaluate the condition regardless of the pipeline failure.
2) this could be a "hook_on_skipped_solid" which gets triggered when there's an execution skipped event.

**Comment by AmirAktify** [comment]
Hello! 
It's mostly like Case # 1 you refer to.

To give some more specifics, suppose you have solids A_1 through A_n. Let's say B is dependent / waiting on all of A_1 through A_n. Once they all finish (fail or succeed), B wants to run some cleanup process + other pipeline business logic. This cleanup process cannot be run earlier or linked with each A_i separately, because the cleanup process impacts resource availability for all of the A_i.

Let's say A_3 fails, but the other A_i's succeed. But I'd want a hook to do at least the cleanup logic that B couldn't, because it never ran.

Now of course, one could also modify the pipeline where each A_i catches its own errors internally in some try / except / catch, and completes no matter what. That however removes usage of some of the nice features dagster allows on failures, such as notifications on each failing A_i, retries, etc.

**Comment by yuhan** [comment]
@AmirAktify Thanks for the details! I'm working on https://github.com/dagster-io/dagster/issues/3613 which will likely tackle use case you described.

**Comment by yuhan** [comment]
We introduced [run status sensors](https://docs.dagster.io/concepts/partitions-schedules-sensors/sensors#run-status-sensors) in 0.12.4. It should help in this case that you could have a run status sensors that monitor failed runs and operate from there. 

Alternatively, you could also have a custom sensor which monitors the events through `context.instance.get_event_records` where you could specify the event type that you want to monitor and set up policies accordingly. Example: [asset sensor](https://docs.dagster.io/concepts/partitions-schedules-sensors/sensors#multi-asset-sensors).
